components/tooltip/tooltip.js

```diff
@@ -931,25 +946,49 @@ class Tooltip extends RtlMixin(LitElement) {
 		this.showing = this._isFocusing || this._isHovering || this.forceShow;
 	}
 
-	_updateTarget() {
+	async _updateTarget() {
 		this._removeListeners();
-		this._target = this._findTarget();
+
+		if (this._target && this.id) {
+			elemIdListRemove(this._target, 'aria-labelledby', this.id);
+			elemIdListRemove(this._target, 'aria-describedby', this.id);
+			if (this.#targetDelegated) {
+				this._target.removeAttribute('aria-label');
+				this._target.removeAttribute('aria-description');
+			}
+		}
+
+		this._target = await this._findTarget();
 		if (this._target) {
 			const isInteractive = this._isInteractive(this._target);
 			this.id = this.id || getUniqueId();
 			this.setAttribute('role', 'tooltip');
-			if (this.forType === 'label') {
-				elemIdListAdd(this._target, 'aria-labelledby', this.id);
-			} else if (!this.announced || isInteractive) {
-				elemIdListAdd(this._target, 'aria-describedby', this.id);
-			}
-			if (logAccessibilityWarning && !isInteractive && !this.announced) {
-				console.warn(
-					'd2l-tooltip may be being used in a non-accessible manner; it should be attached to interactive elements like \'a\', \'button\',' +
-					'\'input\'', '\'select\', \'textarea\' or static / custom elements if a role has been set and the element is focusable.',
+
+			if (this.#targetDelegated) {
+				if (this.forType === 'label') {
+					this.removeAttribute('aria-labelledby');
+					this._target.ariaLabel = this.shadowRoot.querySelector('slot').assignedNodes().map(n => n.textContent).join('\n');
+				} else if (!this.announced || isInteractive) {
+					this.removeAttribute('aria-describedby');
+					this._target.ariaDescription = this.shadowRoot.querySelector('slot').assignedNodes().map(n => n.textContent).join('\n');
+				}
+			}
+			else {
+				if (this.forType === 'label') {
+					this._target.removeAttribute('aria-label');
+					elemIdListAdd(this._target, 'aria-labelledby', this.id);
+				} else if (!this.announced || isInteractive) {
+					this._target.removeAttribute('aria-description');
+					elemIdListAdd(this._target, 'aria-describedby', this.id);
+				}
```

Comment:
Removing the `label`/`description` isn't strictly necessary here, but they won't be used since the `...by` property we're setting takes precedence, and cleaning them up seems good.

Reply:
So if I'm reading this correctly, there's no way for the downstream code (e.g. `d2l-button`) to leverage `aria-labelledby`/`aria-decribedby` to set a label/description themselves on their `<button>` and then have a consumer point a tooltip at them... since that would fall into the `targetDelegated` case which would then remove their label/description?

The only way it could work would be if `<d2l-button>` itself rendered the tooltip, in which case their in the same scope and therefore _not_ delegated so the tooltip `id` would get added to the list.

Reply:
Right, I couldn't find a convenient way to get the computed `label`/`description`, besides by just getting the element via the id and grabbing the text content, which seems like it's just as well to handle once in the tooltip code rather than duplicating that in every delegated-focus component.

Obviously we will someday be able to use the ElementInternals/AOM stuff, but until then this seems better than having nothing.

> The only way it could work would be if <d2l-button> itself rendered the tooltip, in which case they're in the same scope and therefore not delegated so the tooltip id would get added to the list.

Yes, that's exactly right. I think there could be some much more complicated solutions, but I'm not sure it's worth the effort.

Reply:
Cool, makes sense! I thought of an alternate approach last night... so as a thought experiment:

What if the components themselves opted into this using something like a `TooltipTargetMixin`? That mixin could come with some reactive state properties like `tooltipValue` and `tooltipType`, and the tooltip code would set those whenever the tooltip value changed or it was shown/hidden.

That would then put the responsibility on the target to figure out what to actually do with the value and what element to put it on -- using `aria-label/description`, maybe an offscreen element hooked up with `aria-labelledby` if there are multiple labels, etc.

The advantage to this approach is that it empowers the target element to "do what's best" and hooks into Lit's lifecycle to do the rendering. Tooltip could even `throw` if the target didn't implement the mixin. It's obviously more work though for each one to opt in.

My worry with the more "automatic" approach is that consumers can accidentally mess up a core component's accessibility by unknowingly completely overriding its label. But maybe that's what we want anyway... if you're pointing a tooltip at a button or whatever, you probably want its value to become the label. 🤷

Reply:
This overlaps with the comments I just made.
https://github.com/BrightspaceUI/core/pull/4606/files#r1571060598
https://github.com/BrightspaceUI/core/pull/4606/files#r1571072507

Dave, if I follow you correctly, that would address my concern about modifying the DOM outside of the target custom element's `render` method. That's where my mind was going too, but you put a bit more thought into the sol'n. 😄

Reply:
Yeah, that was my intent.

Reply:
Yeah, I agree with all of this. I'll investigate a little more, and try to get some numbers on how many places would need to be updated.